### ISSUE TITLE
What if there is no masking layer but I use 0-padded sequences? (OR Why use Masking Layer?)

### ISSUE THREAD

**Issue by shyamupa** [fchollet/keras#1855]
This is a question to understand what exactly is the purpose of Masking layer. I have a complicated model in which I have input sequences of different lengths. I am 0-padding using the `pad_sequences` method in `keras.preprocessing.sequence`. The problem is my model is not learning at all (loss remains same), and I suspect 0-padding is the problem. 

I tried adding a masking layer just after the input and before the embedding layer using `model.add_node(Masking(), name='masked_input', input='input')` as good measure, but this lead to compilation errors. I am using things like `TimeDistributedDense` so I suspect I will not be able to use masking.

*Shouldn't 0-padding be enough? Why exactly do I need to have a masking layer?*

Please make sure that the boxes below are checked before you submit your issue. Thank you!

- [x] Check that you are up-to-date with the master branch of Keras. You can update with:
pip install git+git://github.com/fchollet/keras.git --upgrade --no-deps

- [x] If running on Theano, check that you are up-to-date with the master branch of Theano. You can update with:
pip install git+git://github.com/Theano/Theano.git --upgrade --no-deps

- [ ] Provide a link to a GitHub Gist of a Python script that can reproduce your issue (or just copy the script here if it is short).

**Comment by pasky** [comment]
I guess the first thing might be to try to instead chop your sequences to have typically minimum or no zero padding and some part missing, rather than padding to accommodate the longest sequence.  What you describe (no loss changes at all) sounds strange to me, and I'd recommend more direct ways of debugging first (simplifying the model, changing layer sizes, ...).

(I'm no expert in masking though, since CNNs don't support it and my models often have a CNN on top of an RNN or the input sequence.)

**Comment by ipod825** [comment]
Don't know if you have exactly the same problem, but this might help #1800

**Comment by shyamupa** [comment]
@ipod825 Thanks a lot for the pointer! I will try this and see if it helps.
@pasky - Thanks for the tip. I tried truncating and also simplifying the model (removed attention layer), but still nothing. Pretty sure now its a silly mistake.

**Comment by LeZhengThu** [comment]
@shyamupa  Hi, could you tell me what's your problem or mistake? I'm having the same issue and don't know where to find the problem. Thanks.
